src/core/net/tcp6.cpp

```diff
@@ -302,27 +497,271 @@ Error Tcp::Listener::Deinitialize(void)
+        if (nextAction != RELOOKUP_REQUIRED)
+        {
+            ProcessSignals(*endpoint, priorHead, sig);
+            ExitNow();
+        }
+        /* If the matching socket was in the TIME-WAIT state, then we try passive sockets. */
+    }
+
+    listener = mListeners.FindMatching(aMessageInfo, listenerPrev);
+    if (listener != nullptr)
+    {
+        struct tcpcb_listen *tpl = &listener->GetTcbListen();
+
+        tcp_input(ip6Header, tcpHeader, &aMessage, nullptr, tpl, nullptr);
+        ExitNow();
+    }
+
+    tcp_dropwithreset(ip6Header, tcpHeader, nullptr, &InstanceLocator::GetInstance(), length - headerSize,
+                      ECONNREFUSED);
+
+exit:
+    return error;
+}
+
+void Tcp::ProcessSignals(Endpoint &aEndpoint, otLinkedBuffer *aPriorHead, struct signals &aSignals)
+{
+    if (!IsInitialized(aEndpoint) || aEndpoint.IsClosed())
+    {
+        ExitNow();
+    }
+    if (aEndpoint.mSendDoneCallback != nullptr)
+    {
+        otLinkedBuffer *curr = aPriorHead;
+        for (int i = 0; i != aSignals.links_popped; i++)
+        {
+            otLinkedBuffer *next = curr->mNext;
+
+            if (i != 0 && (!IsInitialized(aEndpoint) || aEndpoint.IsClosed()))
+            {
+                ExitNow();
+            }
+
+            curr->mNext = nullptr;
+            aEndpoint.mSendDoneCallback(&aEndpoint, curr);
+            curr = next;
+        }
+    }
+
+    if (!IsInitialized(aEndpoint) || aEndpoint.IsClosed())
+    {
+        ExitNow();
+    }
+    if (aSignals.conn_established && aEndpoint.mEstablishedCallback != nullptr)
+    {
+        aEndpoint.mEstablishedCallback(&aEndpoint);
+    }
+
+    if (!IsInitialized(aEndpoint) || aEndpoint.IsClosed())
     {
-        OT_UNUSED_VARIABLE(active);
+        ExitNow();
+    }
+    if ((aSignals.recvbuf_notempty || aSignals.rcvd_fin) && aEndpoint.mReceiveAvailableCallback != nullptr)
+    {
+        aEndpoint.mReceiveAvailableCallback(&aEndpoint, cbuf_used_space(&aEndpoint.GetTcb().recvbuf),
+                                            aEndpoint.GetTcb().reass_fin_index != -1,
+                                            cbuf_free_space(&aEndpoint.GetTcb().recvbuf));
     }
 
-    for (Listener &passive : mListeners)
+    if (!IsInitialized(aEndpoint) || aEndpoint.IsClosed())
     {
-        OT_UNUSED_VARIABLE(passive);
+        ExitNow();
+    }
+    if (aEndpoint.GetTcb().t_state == TCP6S_TIME_WAIT && aEndpoint.mDisconnectedCallback != nullptr)
+    {
+        aEndpoint.mDisconnectedCallback(&aEndpoint, OT_TCP_DISCONNECTED_REASON_TIME_WAIT);
+    }
+
+exit:
+    return;
+}
+
+Error Tcp::BsdErrorToOtError(int aBsdError)
+{
+    Error error = kErrorFailed;
+
+    switch (aBsdError)
+    {
+    case 0:
+        error = kErrorNone;
+        break;
     }
 
     return error;
 }
 
+bool Tcp::CanBind(const SockAddr &aSockName)
+{
+    uint16_t port    = HostSwap16(aSockName.mPort);
+    bool     allowed = true;
```

Comment:
nit: I think we could reduce a bit of code by:

- Initialize `allowed` to `false`.
- Remove `allowed = false` from `ExitNow()` instances below.
- Add `allowed = true;` before the `exit:` label.

Thoughts?